niceman/resource/ssh.py

```diff
@@ -0,0 +1,189 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the niceman package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Resource sub-class to provide management of a SSH connection."""
+
+import attr
+import uuid
+from pipes import quote
+
+import logging
+lgr = logging.getLogger('niceman.resource.ssh')
+
+from .base import Resource, attrib
+from ..support.starcluster.sshutils import SSHClient
+
+
+@attr.s
+class SSH(Resource):
+
+    # Generic properties of any Resource
+    name = attr.ib()
+
+    # Configurable options for each "instance"
+    host = attrib(doc="DNS or IP address of server")
+    port = attrib(default=22,
+        doc="Port to connect to on remote host")
+    key_filename = attrib(default=None,
+        doc="Path to SSH private key file matched with AWS key name parameter")
+    user = attrib(default=None,
+        doc="Username to use to log into remote environment")
+    password = attrib(default=None,
+        doc="Password to use to log into remote environment")
+
+    id = attr.ib(default=None)  # EC2 instance ID
+
+    type = attr.ib(default='ssh')  # Resource type
+
+    # Current instance properties, to be set by us, not augmented by user
+    status = attr.ib(default=None)
+    _ssh = attr.ib(default=None)
+
+    def connect(self):
+        """
+        Open a connection to the environment resource.
+        """
+        self._ssh = SSHClient(
+            self.host,
+            username=self.user,
+            password=self.password,
+            private_key=self.key_filename,
+            port=int(self.port)
+        )
+
+    def create(self):
+        """
+        Register the SSH connection to the niceman inventory registry.
+
+        Returns
+        -------
+        dict : config and state parameters to capture in the inventory file
+        """
+        self.id = str(uuid.uuid4())
+        self.status = 'N/A'
+        return {
+            'id': self.id,
+            'status': self.status,
+            'host': self.host,
+            'user': self.user,
+            'port': self.port,
+            'key_filename': self.key_filename,
+        }
+
+    def delete(self):
+        return
+
+    def start(self):
+        return
+
+    def stop(self):
+        return
+
+    def get_session(self, pty=False, shared=None):
+        """
+        Log into remote environment and get the command line
+        """
+        assert self._ssh, "We should create or connect to remote server first"
+
+        return (PTYSSHSession if pty else SSHSession)(
+            ssh=self._ssh
+        )
+
+# Alias SSH class so that it can be discovered by the ResourceManager.
```

Comment:
any why otherwise it (`SSH`) is not discovered?

Reply:
Right, it will not find the module because the discovery process is looking for CamelCase class names based on the hyphen-delineated resource types declared.

e.g. docker-container looks for class DockerContainer

likewise, a resource of type "ssh" is looking for a class "Ssh"